File: app/src/main/java/tw/tonyyang/englishwords/WordListM2Adapter.java
package tw.tonyyang.englishwords;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import java.util.List;

import tw.tonyyang.englishwords.database.Word;


/**
 * Created by tonyyang on 2017/6/3.
 */

public class WordListM2Adapter extends RecyclerView.Adapter<WordListM2Adapter.ViewHolder> {

    interface OnRecyclerViewListener {
        void onItemClick(View v, int position);

        void onItemLongClick(View v, int position);
    }

    private List<Word> wordList;

    private OnRecyclerViewListener onRecyclerViewListener;

    void setOnRecyclerViewListener(OnRecyclerViewListener onRecyclerViewListener) {
        this.onRecyclerViewListener = onRecyclerViewListener;
    }

    WordListM2Adapter(List<Word> wordList) {
        this.wordList = wordList;
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(ViewGroup viewGroup, int i) {
        View view = LayoutInflater.from(viewGroup.getContext()).inflate(R.layout.word_list_m2_item, viewGroup, false);
        ViewHolder vh = new ViewHolder(view);
        view.setOnClickListener(vh);
        return vh;
    }

    @Override
    public void onBindViewHolder(ViewHolder viewHolder, int i) {
        Word words = wordList.get(i);
        viewHolder.image.setImageResource(R.drawable.book);
        viewHolder.title.setText(words.getWord());
        viewHolder.info.setText(words.getWordMean());
        viewHolder.category.setText(words.getCategory());
    }

    @Override
    public int getItemCount() {
        return wordList.size();
    }

    Word getItem(int position) {
        return wordList.get(position);
    }

    class ViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener, View.OnLongClickListener {

        private ImageView image;
        private TextView title;
        private TextView info;
        private TextView category;

        ViewHolder(View itemView) {
            super(itemView);
            image = itemView.findViewById(R.id.image);
            title = itemView.findViewById(R.id.title);
            info = itemView.findViewById(R.id.info);
            category = itemView.findViewById(R.id.category);
        }

        @Override
        public void onClick(View v) {
            if (onRecyclerViewListener != null) {
                onRecyclerViewListener.onItemClick(v, getAdapterPosition());
            }
        }

        @Override
        public boolean onLongClick(View v) {
            if (onRecyclerViewListener != null) {
                onRecyclerViewListener.onItemLongClick(v, getAdapterPosition());
            }
            return false;
        }
    }
}

File: app/src/main/java/tw/tonyyang/englishwords/WordListM2Adapter.kt
package tw.tonyyang.englishwords

import android.view.LayoutInflater
import android.view.View
import android.view.View.OnLongClickListener
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import tw.tonyyang.englishwords.database.Word

class WordListM2Adapter internal constructor(private val wordList: List<Word>) : RecyclerView.Adapter<WordListM2Adapter.ViewHolder>() {
    interface OnRecyclerViewListener {
        fun onItemClick(v: View?, position: Int)
        fun onItemLongClick(v: View?, position: Int)
    }

    private var onRecyclerViewListener: OnRecyclerViewListener? = null
    fun setOnRecyclerViewListener(onRecyclerViewListener: OnRecyclerViewListener?) {
        this.onRecyclerViewListener = onRecyclerViewListener
    }

    override fun onCreateViewHolder(viewGroup: ViewGroup, i: Int): ViewHolder {
        val view = LayoutInflater.from(viewGroup.context).inflate(R.layout.word_list_m2_item, viewGroup, false)
        val vh = ViewHolder(view)
        view.setOnClickListener(vh)
        return vh
    }

    override fun onBindViewHolder(viewHolder: ViewHolder, i: Int) {
        val words = wordList[i]
        viewHolder.image.setImageResource(R.drawable.book)
        viewHolder.title.text = words.word
        viewHolder.info.text = words.wordMean
        viewHolder.category.text = words.category
    }

    override fun getItemCount(): Int {
        return wordList.size
    }

    fun getItem(position: Int): Word {
        return wordList[position]
    }

    inner class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView), View.OnClickListener, OnLongClickListener {
        val image: ImageView = itemView.findViewById(R.id.iv_image)
        val title: TextView = itemView.findViewById(R.id.tv_title)
        val info: TextView = itemView.findViewById(R.id.tv_info)
        val category: TextView = itemView.findViewById(R.id.tv_category)
        override fun onClick(v: View) {
            onRecyclerViewListener?.onItemClick(v, adapterPosition)
        }

        override fun onLongClick(v: View): Boolean {
            onRecyclerViewListener?.onItemLongClick(v, adapterPosition)
            return false
        }
    }
}